s64da_benchmark_toolkit/reporting.py

```diff
@@ -0,0 +1,170 @@
+                self._save_query_output(query_metric)
+
+            self.df = self.df.append(query_metric.dataframe)
+
+        self.df = self.df.reset_index(drop=True)
+
+        netdata_config = self.config.get('netdata')
+        if netdata_config:
+            netdata = Netdata(netdata_config)
+            if self.netdata_output_file:
+                netdata.write_stats(self.df, self.netdata_output_file)
+
+        if self.scale_factor:
+            self._check_correctness()
+
+        total_runtime = self.df['timestamp_stop'].max() - self.df['timestamp_start'].min()
+        self.total_runtime_seconds = total_runtime.total_seconds()
+        self._print_results()
+
+        print(f'\nTotal runtime: {total_runtime} ({self.total_runtime_seconds:.2f}s)')
+
+    def _save_explain_plan(self, query_metric):
+        plan_file_name = query_metric.make_file_name('txt')
+        plan_file_path = os.path.join(self.explain_analyze_dir, plan_file_name)
+        os.makedirs(os.path.dirname(plan_file_path), exist_ok=True)
+
+        with open(plan_file_path, 'w') as plan_file:
+            plan_file.write(query_metric.plan)
+
+    def _save_query_output(self, query_metric):
+        query_result = query_metric.result
+        if query_result is not None and query_result[1]:
+            query_result_header = query_result[0]
+            query_result_data = query_result[1]
+
+        else:
+            query_result_header = []
+            query_result_data = []
+
+        csv_file_name = query_metric.make_file_name('csv')
+        csv_file_path = os.path.join(self.query_results, csv_file_name)
+        os.makedirs(os.path.dirname(csv_file_path), exist_ok=True)
+        with open(csv_file_path, 'w') as csv_file:
+            csvfile = csv_writer(csv_file)
+            csvfile.writerow(query_result_header)
+            csvfile.writerows(query_result_data)
+
+    def _sort_df(self):
+        index_sort = index_natsorted(zip(self.df['stream_id'], self.df['query_id']))
+        self.df = self.df.reindex(index=order_by_index(self.df.index, index_sort))
+        self.df = self.df.reset_index(drop=True)
+
+    def _print_results(self):
+        self._sort_df()
+
+        if 'print' in self.output:
+            print('\n')
+            print(tabulate(self.df, headers='keys', tablefmt='github', floatfmt='.2f'))
+
+        if 'csv' in self.output:
+            if self.csv_file:
+                self.df.to_csv(self.csv_file, sep=';')
+
+        if 'correctness_check' in self.df:
+            report_datetime = datetime.now().strftime('%d-%m-%Y %H:%M:%S')
+            with open(self.html_output, 'w') as html_output_file:
+                html_output_file.write(f'''
+<h1>Swarm64 Benchmark Results Report</h1>
+<p>{ report_datetime }</p>
+<p>Total runtime: { self.total_runtime_seconds }s</p>\n''')
+                self.df.to_html(buf=html_output_file, escape=False, formatters={
+                    'correctness_check': CorrectnessResult.format_html
```

Comment:
Please consider changing this to: `'correctness_check': lambda v: v.format_html()` and make `format_html` in `correctness.py` a normal class function - then you don't have to supply the value. I think it will make the code easier to follow.

Reply:
Great call. Added.